File: sdks/python/benches/sdk_roundtrip.py
"""Python SDK binding-overhead benchmarks.

Measures the cost of the PyO3 boundary alone — no network. We use
`describe()` (the canonical hot path the recent caching commits 918f77e /
4f21aac targeted) and a constructor microbench. The autoresearch oracle
divides the Python p99 by the Rust-direct p99 to derive `py_overhead_ratio`,
so any binding regression shows up immediately.

No credentials needed: `describe()` works with an empty config because
the manifest is static metadata. This keeps the bench fully reproducible.

Run via `pytest sdks/python/benches/ --benchmark-only --benchmark-json=...`.
"""

from __future__ import annotations

import pytest

from openpx._native import NativeExchange


@pytest.fixture(scope="module")
def kalshi_native() -> NativeExchange:
    # Empty dict is sufficient for describe() — manifest metadata is static.
    return NativeExchange("kalshi", {})


@pytest.fixture(scope="module")
def polymarket_native() -> NativeExchange:
    return NativeExchange("polymarket", {})


@pytest.mark.benchmark(group="describe_cached")
def test_describe_kalshi_cached(benchmark, kalshi_native: NativeExchange) -> None:
    # Prime the OnceLock so we measure the cached fast path the user-facing
    # `describe()` call hits in steady state.
    kalshi_native.describe()
    benchmark(kalshi_native.describe)


@pytest.mark.benchmark(group="describe_cached")
def test_describe_polymarket_cached(benchmark, polymarket_native: NativeExchange) -> None:
    polymarket_native.describe()
    benchmark(polymarket_native.describe)


@pytest.mark.benchmark(group="constructor")
def test_construct_native_kalshi(benchmark) -> None:
    # Cold-path baseline. Constructor cost includes config depythonize +
    # ExchangeInner::new + Arc allocation. Any regression here means the
    # binding got heavier on first contact.
    benchmark(lambda: NativeExchange("kalshi", {}))


@pytest.mark.benchmark(group="constructor")
def test_construct_native_polymarket(benchmark) -> None:
    benchmark(lambda: NativeExchange("polymarket", {}))


@pytest.mark.benchmark(group="getter")
def test_id_getter_kalshi(benchmark, kalshi_native: NativeExchange) -> None:
    # Pure attribute getter — the cheapest path across the boundary.
    # Useful as a floor/sanity reference.
    benchmark(lambda: kalshi_native.id)

File: sdks/typescript/benches/sdk_roundtrip.mjs
// TypeScript SDK binding-overhead benchmarks.
//
// Measures the cost of the NAPI boundary alone — no network. We bench
// `describe()` (the canonical hot path commit 6fdc2aa optimised — bytes
// cache + per-call deserialize) and a constructor microbench. The
// autoresearch oracle divides the Node p99 by the Rust-direct p99 to
// derive `ts_overhead_ratio`.
//
// No credentials needed: `describe()` works with an empty config because
// the manifest is static metadata.
//
// Run via `node --experimental-vm-modules sdks/typescript/benches/sdk_roundtrip.mjs`.

import { Bench } from "tinybench";
import { Exchange } from "../index.js";

const TIME_MS = 2000; // per-bench wall-clock budget

const bench = new Bench({ time: TIME_MS });

const kalshi = new Exchange("kalshi", {});
const polymarket = new Exchange("polymarket", {});

// Prime caches so the steady-state numbers are what we report.
kalshi.describe();
polymarket.describe();

bench
  .add("describe_kalshi_cached", () => {
    kalshi.describe();
  })
  .add("describe_polymarket_cached", () => {
    polymarket.describe();
  })
  .add("construct_kalshi", () => {
    new Exchange("kalshi", {});
  })
  .add("construct_polymarket", () => {
    new Exchange("polymarket", {});
  })
  .add("id_getter_kalshi", () => {
    kalshi.id;
  });

await bench.run();

const tasks = bench.tasks.map((task) => {
  const r = task.result;
  return {
    name: task.name,
    samples: r?.samples?.length ?? 0,
    p50_ns: r?.p50 != null ? Math.round(r.p50 * 1e6) : null,
    p99_ns: r?.p99 != null ? Math.round(r.p99 * 1e6) : null,
    p999_ns: r?.p999 != null ? Math.round(r.p999 * 1e6) : null,
    mean_ns: r?.mean != null ? Math.round(r.mean * 1e6) : null,
    hz: r?.hz ?? null,
  };
});

console.log(JSON.stringify({ time_ms_budget: TIME_MS, tasks }));
